[DataPipe] Fix issue with DataPipe serialization with `dill`

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73119
* #73118
* __->__ #72896

Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

NivekT committed:
[DataPipe] Fix issue with DataPipe serialization with `dill`

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34258669).

NivekT commented on test/test_dataloader.py:
```diff
@@ -1366,6 +1373,29 @@ def test_multiprocessing_contexts(self):
                 self.assertEqual(
                     reference, list(self._get_data_loader(ds_cls(counting_ds_n), multiprocessing_context=ctx, **dl_common_args)))
 
+    def test_multiprocessing_iterdatapipe(self):
```
This test should fail prior to the changes

NivekT commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -27,6 +34,31 @@ def check_lambda_fn(fn):
+    """
+    try:
+        return pickle.dumps(fn), SerializationType("pickle")
+    except (pickle.PickleError, AttributeError):
```
I wonder if I should catch all Exception here.

ejguan commented on torch/utils/data/datapipes/iter/callable.py:
```diff
@@ -109,13 +109,11 @@ def __getstate__(self):
         state = (
             self.datapipe,
-            dill_function,
+            serialized_fn,
+            method,
```
Could we combine these two variables into a tuple for serialization?

NivekT replied:
Done.

ejguan commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -27,6 +34,31 @@ def check_lambda_fn(fn):
         )
 
 
+def serialize_fn(fn, is_dill_available):
```
`is_dill_available` doesn't need to be the argument. We can determine `is_dill_available` inside this function. Then, in terms of the API simplicity, users/developers don't need to provide such static variable for each function.

Besides, we can let this function handle all functions dynamically. We can provide a default `__getstate__`/`__setstate__` here https://github.com/pytorch/pytorch/blob/17b3ba148de5bca4da66b4045e451b245f916179/torch/utils/data/dataset.py#L317 and loop over `__dict__` to invoke this function.

NivekT replied:
I updated to remove `is_dill_available` as an argument, and also created/updated the default `__getstate__`/`__setstate__` for `IterDataPipe` and `MapDataPipe`.

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

There will be a follow-up PR in TorchData as well

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/selecting.py:
```diff
@@ -83,17 +73,4 @@ def _isNonEmpty(self, data):
-            self.filter_fn = dill.loads(dill_function)  # type: ignore[assignment]
-        else:
-            self.filter_fn = dill_function  # type: ignore[assignment]
+        return super().__getstate__()
```
This can be removed.

ejguan commented on torch/utils/data/datapipes/iter/selecting.py:
```diff
@@ -1,22 +1,14 @@
 
 from torch.utils.data import IterDataPipe, functional_datapipe
 from torch.utils.data.datapipes.dataframe import dataframe_wrapper as df_wrapper
+from torch.utils.data.datapipes.utils.common import DILL_AVAILABLE, check_lambda_fn
```
Why `DILL_AVAILABLE` is needed here?

NivekT replied:
Actually they aren't so I am going to remove them now.

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

There will be a follow-up PR in TorchData as well

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -230,32 +231,4 @@ def __iter__(self):
-        else:
-            self.group_key_fn = dill_function  # type: ignore[assignment]
-        self.wrapper_class = DataChunk
+        return super().__getstate__()
```
Removing `__getstate__` should be fine here.

ejguan commented on torch/utils/data/datapipes/iter/callable.py:
```diff
@@ -108,30 +108,7 @@ def __len__(self) -> int:
     def __getstate__(self):
         if IterDataPipe.getstate_hook is not None:
             return IterDataPipe.getstate_hook(self)
```
Ditto.

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34258669).

NivekT commented on torch/utils/data/dataset.py:
```diff
@@ -317,7 +334,20 @@ def class_function(cls, enable_df_api_tracing, source_dp, *args, **kwargs):
-        return self.__dict__
+        state_dict = {}
+        for k, v in self.__dict__.items():
+            if callable(v):
```
I choose to only apply the `serialize_fn` on callable objects here. Any objection to this?

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34258669).

ejguan reviewed (commented):
Can you add a test here https://github.com/pytorch/pytorch/blob/bb736fac3338f1ab8f52b9ff16e6ce91eb20572c/test/test_datapipe.py#L588-L590 to validate if the outputs are the same between original DataPipe and deserialized/serialized DataPipe?

NivekT commented:
> Can you add a test here
> 
> https://github.com/pytorch/pytorch/blob/bb736fac3338f1ab8f52b9ff16e6ce91eb20572c/test/test_datapipe.py#L588-L590
> 
> to validate if the outputs are the same between original DataPipe and deserialized/serialized DataPipe?

This is done

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34258669).

VitalyFedyunin commented on test/test_dataloader.py:
```diff
@@ -62,6 +63,13 @@
+try:
+    import numpy as np
+except ImportError:
+    np = None
```
this is not very helpful, as tests will break anyway, better add skipping mechanism

NivekT replied:
I later check if `np` is available before attempting to use it, and use apply a non-`np`, equivalent transformation instead.

https://github.com/pytorch/pytorch/pull/72896/files#diff-7cd4c7412591dcd9a7f8b7eb6622e286d9c19e3835bc8319af2ab4a1cf2e7badR1380

Other options:
1. Import it within the test (I was worried that will make `np` non-global and accidentally pass a test that should fail
2. Skip when `np` is not available

VitalyFedyunin replied:
But it will correctly work only if np exists, otherwise, even old code will pass this test.
Also because `np` here is global, you are putting landmine under all users of `np` here, as they would have (and sometimes forget) to do same `if np in not None` check.

NivekT replied:
I see, that makes sense. I will skip the test if `np` is missing then.

NivekT replied:
This is done

ejguan commented on torch/utils/data/_utils/serialization.py:
```diff
@@ -0,0 +1,49 @@
+import pickle
```
I assume this file needs to be here to prevent cyclic reference. Otherwise, putting it to https://github.com/pytorch/pytorch/tree/master/torch/utils/data/datapipes/utils makes more sense to me.

If we want to move it into datapipes folder, we may need to completely decouple `IterDataPipe` and `MapDataPipe` from `dataset.py` and maybe put them into `torch/utils/data/datapipes/utils/__init__`

NivekT replied:
Yea, it is there to avoid cyclic reference. We can move it inside `datapipes` folder once `IterDataPipe` and `MapDataPipe` move out of `dataset.py`. 

I will look into that soon.

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34258669).

ejguan commented on test/test_dataloader.py:
```diff
@@ -1366,6 +1375,31 @@ def test_multiprocessing_contexts(self):
+        def row_processer(row):
+            return np.add(row, 1)
+
+        def filter_len(row):
+            return len(row) == 4
```
Local function is not pickable. You need to put functions out of `test_multiprocessing_iterdatapipe`

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34258669).

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34258669).

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fix issue with DataPipe serialization with `dill`"


Fixing the issue described here: https://github.com/pytorch/data/issues/214

The fix involves creating new default methods to serialize/deserialize callable methods within DataPipes. This PR also adds various new tests to ensure that the behavior works as expeced.

There will be a follow-up PR in TorchData as well.

Differential Revision: [D34258669](https://our.internmc.facebook.com/intern/diff/D34258669)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34258669).
